### ISSUE TITLE
Feedback on generated code

### ISSUE THREAD

**Issue by vfdev-5** [pytorch-ignite/trainite#74]
- [x] 1) Instead of https://github.com/pytorch-ignite/trainite/blob/bf6f4dfaf91f133ab0219777707454905e92c25f/trainite/shared/main.py#L162-L168
let's use ignite's https://github.com/pytorch/ignite/blob/a9c793e2fe8976f0bd92ced798f74251a1d74554/ignite/utils.py#L239

- [x] 2) Simplify device specification using `idist.device()`: 
Define as following 
```diff
class ProjectConfig(BaseModel):
    ...
    - device: str = "auto"
    + device: str | None = None
```
and in the code, remove `resolve_device` and do : 
```python
device = idist.device() if config.device is None else config.device
``` 

- [x] 3) If we have trainer abstraction I do not understand why we now put into shared/main.py all language model specific things.
We can make a single file specific code instead of two. 
We can make main.py file generic by creating the trainer. Trainer class should be defined in the config like the model or the optimizer and in main.py 
```python
def main() -> None:
    args = parse_args()
    config_path = Path(args.config)
    config = load_config(config_path, ProjectConfig)

    # Later we can insert the code for DDP: with idist.Parallel(...)
    trainer = instantiate(config.trainer, config=config)
    trainer.run()

if __name__ == "__main__":
    main()    
```

Let's decoder instantiate everything in the constructor in few steps:
- dataflow
- model, optimizer, lr scheduler, 
- engine and handlers

- [x] 4) Let's remove `class ComponentConfig(BaseModel):` and each Config class should inherit from pydantic BaseModel. No subclassing.

- [x] 5) Apply "separate concerns" logic in for dataflow creation. Instead of 
```python
    tokenizer = instantiate(config.preprocessor) if config.preprocessor is not None else None

    train_loader, val_loader, test_loader = build_dataloaders(config)

    vocab_size = resolve_vocab_size(tokenizer, config.model)  # <-- I do not think model config should contain vocab_size, we take it from the tokenizer.
    model = build_model(config.model, tokenizer, vocab_size, device)

    optimizer = instantiate(config.optimizer, params=model.parameters())

    # The transform owns the prompt format; hand it to the trainer for inference logging.
    ds = train_loader.dataset
    ds = ds.dataset if isinstance(ds, Subset) else ds
    prompt_transform = getattr(ds, "transform", None)
```
let's do the following:
```python
# This is the trainer constructor:
    self.train_loader, self.val_loader, self.test_loader, self.tokenizer = build_dataloaders(config)
    self.model = build_model(config.model, tokenizer, device)
    self.optimizer = instantiate(config.optimizer, params=model.parameters())
```

**Comment by vfdev-5** [comment]
Few more feedback:
- [x] 1 self.trainer_config shortcut for self.config.trainer

- [x] 2 vocab_size is only in tokenizer, remove resolve_vocab_size and simplify build_model

- [x] 3 loss_fn -> criterion

- [x] 4 self.engine -> self.trainer

~5 self.train_evaluator and self.test_evaluator -> remove self.test_evaluator and rename self.train_evaluator -> self.extra_evaluator~

- [x] 6 let's make a subset of training set of the size of validation set. -> build_dataloaders -> train_loader, val_loader, test_loader

```python
self.extra_evaluator.run(self.train_loader, max_iters=len(self.val_loader))
```

- [x] 7 let's split setup_checkpointing method into two: setup_training_checkpointing() and setup_best_model_checkpoint()

- [x] 8 setup_experiment_logger -> setup_experiment_tracking

- [x] 9 move engine.attach _run_evaluations to the top of the ctor

- [x] 10 _flatten as a separate function in the bottom of the file

- [x] 11 Trainer.run() should be just after the _eval_step

- [x] 12 Save the config yml to output dir

**Comment by vfdev-5** [comment]
@aaishwarymishra @TahaZahid05 can close this one?

**Comment by aaishwarymishra** [comment]
Yes it's all done:)
